pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/BaseOffHeapMutableDictionary.java

```diff
@@ -0,0 +1,411 @@
+      final int offsetInBuf = (hashVal & mask) * NUM_COLUMNS;
+      boolean done= false;
+      for (int i = offsetInBuf; i < offsetInBuf + NUM_COLUMNS; i++) {
+        if (iBuf.get(i) == NULL_VALUE_INDEX) {
+          iBuf.put(i, entry.getValue());
+          done = true;
+          break;
+        }
+      }
+      if (!done) {
+        newOverflowMap.put(entry.getKey(), entry.getValue());
+      }
+    }
+    ValueToDictId newValueToDictId = new ValueToDictId(newList, newOverflowMap);
+    _valueToDict.set(newValueToDictId);
+
+    // We should not clear oldOverflowMap or oldList here, as readers may still be accessing those.
+    // We let GC take care of those elements.
+    return iBuf;
+  }
+
+  private IntBuffer expand() {
+    ValueToDictId valueToDictId = _valueToDict.get();
+    List<IntBuffer> iBufList = valueToDictId.getIBufList();
+    final int numBuffers = iBufList.size();
+    if (numBuffers == 0) {
+      return expand(_sizeOfFirstBuf);
+    }
+    final int lastCapacity = iBufList.get(numBuffers-1).capacity();
+    int expansionMultiple = EXPANSION_MULTIPLES[EXPANSION_MULTIPLES.length-1];
+    if (numBuffers < EXPANSION_MULTIPLES.length) {
+      expansionMultiple = EXPANSION_MULTIPLES[numBuffers];
+    }
+    int newSize = validateSize(lastCapacity, expansionMultiple);
+    return expand(newSize);
+  }
+
+  private int validateSize(int lastCapacity, int expansionMultiple) {
+    long newSize = (long) lastCapacity * expansionMultiple;
+    while (newSize >= Integer.MAX_VALUE) {
+      newSize = newSize/2;
+    }
+    return (int)newSize;
+  }
+
+  protected int nearestPowerOf2(int num) {
+    if ((num & (num -1)) == 0) {
+      return num;
+    }
+    int power = Integer.SIZE - Integer.numberOfLeadingZeros(num);
+    Preconditions.checkState(power < Integer.SIZE-1);
+    return 1 << power;
+  }
+
+  protected int getDictId(@Nonnull Object rawValue) {
+    final int hashVal = Math.abs(rawValue.hashCode());
+    final ValueToDictId valueToDictId = _valueToDict.get();
+    final List<IntBuffer> iBufList = valueToDictId.getIBufList();
+    for (IntBuffer iBuf : iBufList) {
+      final int mask = iBuf.capacity()/NUM_COLUMNS - 1;
+      final int offsetInBuf = (hashVal & mask) * NUM_COLUMNS;
+      for (int i = offsetInBuf; i < offsetInBuf + NUM_COLUMNS; i++) {
+        int dictId = iBuf.get(i);
+        if (dictId != NULL_VALUE_INDEX) {
+          if (rawValue.equals(get(dictId))) {
+            return dictId;
+          }
+        }
+      }
+    }
+    Integer dictId = valueToDictId.getOverflowMap().get(rawValue);
+    if (dictId == null) {
+      return NULL_VALUE_INDEX;
+    }
+    return dictId;
+  }
+
+  protected void indexValue(@Nonnull Object value) {
+    final int hashVal = Math.abs(value.hashCode());
+    ValueToDictId valueToDictId = _valueToDict.get();
+    final List<IntBuffer> iBufList = valueToDictId.getIBufList();
+
+    for (IntBuffer iBuf : iBufList) {
```

Comment:
I am probably reading this incorrectly, but seems that the number of iBuf's ~= cardinality? If so, is this a linear search?